### ISSUE TITLE
Stop creating files in bin directory

### ISSUE THREAD

**Issue by NoICE** [redhat-developer/vscode-java#634]
Random files begin to appear in my bin directory after I generated a lot of code by using my custom code generator. I spent a lot of time debugging where these files are being created, who and why makes them, and I don't want my project polluted with completely useless copies of the project files.

##### Environment
- Operating System: macOS 10.13
- JDK version: multiple
- Visual Studio Code version: Version 1.27.1 (1.27.1)
- Java extension version: 0.30.0

##### Steps To Reproduce
1. Open any kind of project, even non-java ones (mixed), click on any .java file
2. Watch .class files being generated in your bin directory, kotlin files being copied there as-is (which doesn't make any sense), other resources being copied there without any reason.

##### Current Result

Many files are copied to ./bin/ without any reason to do so.

##### Expected Result

No files are created, compiled, copied without developer's consent!

##### Additional Informations

The extension uses jdt_ws which creates these files. I don't want these files to be there, because:

1) I use docker.
2) I don't use vscode for java, I use it for other parts of the project.
3) I compile the java and kotlin files using gradle, to build folder, with multiple other build configuration options not reflected by vscode-java or jdt_ws.
4) The generated files are useless.
5) This extra compilation step burdens battery life and SSD wear.
6) It just doesn't make any sense.

I didn't even know that I have vscode-java installed and it took a LOT of digging to find out WHY and WHO creates these files!

Please consider removing them altogether, OR make this extra compilation step optional (opt-in) with explicit configuration where these files should be created.

If I would change the target directory to "build" or "target", then vs code would be randomly overwriting files compiled by docker/IDEA/gradle. 

They might even be compiled with different version of Java (and JDK)!

**Comment by NoICE** [comment]
Don't get me wrong. I'm glad that this extension exists, it surely looks useful. But it was activated by default and this default "bin behavior" is totally unexpected and uwanted.

**Comment by akaroml** [comment]
@eskibear please take a look as a part of the effort of handling temp files.

**Comment by nik9000** [comment]
I've been using this extension for about 9 months full time. A week or two ago these `bin` directories started popping up. We [usually](https://github.com/elastic/elasticsearch/blob/master/build.gradle#L463-L478) move all of Eclipse's gradle stuff over to a directory that we've gitignored. I'm not sure what has broken, but these `bin` directories are overwriting directories where we put things and breaking normal gradle runs. At this point the plugin I just have to uninstall the plugin.

**Comment by nik9000** [comment]
> A week or two ago

I suspect me seeing this issue late might have something to do with not upgrading very quickly. I'm not sure though.

**Comment by tonny008** [comment]
right click pom.xml and update project configuration

this will generate the .project .classpath .settings files. it seems that the latest version will only generate .project by default. the class file will output to bin folder if there's no .classpath file

**Comment by ewired** [comment]
I solved this issue on Ubuntu by editing the `.classpath` XML file, replacing all `bin/` with `build/temp/` (just to use an example that was already gitignored for my convenience). Before doing that, having this extension enabled would make complete copies of Kotlin source files from the `src/` folder to the `bin/` folder upon saving. Oddly enough, after changing the `.classpath` file, it copies the folder structure and no Kotlin source files into `build/temp/`.

I'm a complete beginner to the Java ecosystem, so I don't know what effects this has long-term, but it works for me. Maybe it's worth exploring automatically editing that file to ensure that it no longer tampers with `bin/`.

The behavior on my setup contradicts tonny008's experience; I think this is due to using Kotlin and the Kotlin Language Server extension which has its own set of issues regarding dependency resolution. Upon starting VS Code, the plugin would re-create the `.classpath`, `.project` files and `.settings` folder, and also pollute the `bin/` folder, and I am running the latest extension version. Even while the `.classpath` file was present and unmodified, it would still pollute the bin folder. As described previously, I edited the `.classpath` file in an attempt to make it pollute a different folder, and it worked.

**Comment by ckuai** [comment]
one more facing this problem. I found whenever I did "mvn eclipse:clean" "mvn eclipse:eclipse" to regenerate the .project and .classpath, then vscode will create a bin folder with all my project's sub folders and files moved under a new created bin folder and build output. Deleting the .factorypath file also has this effect. also my .classpath file has no bin/ path at all. output attributes in .classpath are output="target/test-classes" or output="target/classes"

**Comment by Unichron** [comment]
Same issue here. From observation I think the main problem may be that when you don't have a .classpath, the plugin doesn't import the pom.xml/build.gradle first, but starts building the class files to the default output directory, which happens to be bin.

**Comment by Polish-Civil** [comment]
Making it configurable would be great thing todo

**Comment by testforstephen** [comment]
Reproduced the issue.
1. Use VS Code to open a maven project.
2. Wait the Java Language Server is ready. (:thumbsup: appears at the right bottom status bar)
3. Manually delete `.classpath` file, then a new `bin` folder is automatically created at the root folder, and all non java files are copied there.

Looks like if the .classpath doesn't exists, the Java projects loaded by jdt.ls will take `bin` as the default output folder. And the build process will take any files it doesn't know how to compile as resources, and copy them to the output folder.

**Comment by maxiwu** [comment]
Same issue. I don't understand. Maven project do not use anything from \bin.
How do I avoid this from happening? it messes up my git.

**Comment by donraj** [comment]
Any update on this issue? Seems like my scons build rules are finding multiple copies of themselves and conflicting!

**Comment by donraj** [comment]
I think this is a serious issue, and unfortunately the only solution for me right now is to use idea. sorry VSCODE-JAVA

**Comment by keefbaker** [comment]
Agreed. Until this stops, this extension is basically unusable.

**Comment by janbodnar** [comment]
This is gross. Removed the extension immediately.

**Comment by shua** [comment]
Had to make `.classpath` read-only, because vscode kept overwriting the changed output directories with `bin`. Not a fan.

**Comment by acohen716** [comment]
Any update on this almost 1 year old issue? I've resorted to @ewired [suggestion above](https://github.com/redhat-developer/vscode-java/issues/634#issuecomment-450518553)

**Comment by talios** [comment]
I get this often and it's annoying as (#*$&#(# - esp. since it also copies .git folders into bin which REALLY mess up some tooling.

**Comment by talios** [comment]
@acohen716 the problem with @ewired's "solution" is you'd need to commit those .classpath files for those changes to persist. That's a worse sin in my book.

**Comment by Eskibear** [comment]
> take bin as the default output folder. And the build process will take any files it doesn't know how to compile as resources, and copy them to the output folder.

Based on hypothesis mentioned in https://github.com/redhat-developer/vscode-java/issues/634#issuecomment-489916039 above, I'm wondering if it would be a good solution that:
- If language server doesn't know how to compile as resources, **stop doing further unexpected actions**, simply notify the client (VSCode) about it.
- Then the client can prompt with the error, providing users a shortcut/guidance to fix the issues (e.g. to re-generate a .classpath file)

**Comment by Eskibear** [comment]
Here's another scenario blocked by the issue. I was working on a nodejs project in my workspace, then I dragged a `HelloWorld.java` in, pressed F5, and it just stuck on "compile workspace" for a long time (making no sense for the single .java file). Soon I realized it was copying the huge `node_modules` folder into the `jdt_ws/<workspace>/bin` folder. 

Just FYI that this issue might be more severe than I thought before, especially for polyglot programmers.

**Comment by nemausus** [comment]
Extension is very promising but this issue is a big blocker for me as well.

**Comment by csurfer** [comment]
I was quite hopeful of VSCode's remote ability. Since most of my work is on Java + Python it would have been the ideal solution. This aspect of unwanted files in the bin directory is an issue for me as well.

**Comment by testforstephen** [comment]
> Here's another scenario blocked by the issue. I was working on a nodejs project in my workspace, then I dragged a `HelloWorld.java` in, pressed F5, and it just stuck on "compile workspace" for a long time (making no sense for the single .java file). Soon I realized it was copying the huge `node_modules` folder into the `jdt_ws/<workspace>/bin` folder.
> 
> Just FYI that this issue might be more severe than I thought before, especially for polyglot programmers.

Currently JavaBuilder provides an option `CORE_JAVA_BUILD_RESOURCE_COPY_FILTER
` to exclude the specified resources. It supports a file name pattern (* and ? wild-cards allowed) or the name of a folder which ends with '/'.  For polyglot projects, you need list all possibles to exclude, it will be a burden. 

Another idea is to add a new option (e.g. CORE_JAVA_BUILD_NO_RESOURCE_COPY) in the jdt core, if it's enabled, stop the copy behavior.

**Comment by donraj** [comment]
Any updates @snjeza ? Thanks.

**Comment by hornta** [comment]
Please stop cluttering my workspace with bin folder

**Comment by shua** [comment]
@donraj check out https://github.com/eclipse/eclipse.jdt.ls/pull/1108 for updates on fix

**Comment by esr360** [comment]
Please fix this, it's very frustrating having my text editor create random files for me polluting my source control information - this is going to cause me to accidentally commit rogue files. Very dangerous, why does this happen?

**Comment by snjeza** [comment]
> Please fix this, it's very frustrating having my text editor create random files for me polluting my source control information - this is going to cause me to accidentally commit rogue files. Very dangerous, why does this happen?

@esr360 could you provide a project example (or the  .project, .classpath, pom.xml, build.gradle files)?

**Comment by esr360** [comment]
@snjeza I'm not too sure exactly what you're after but within my project I can see a `pom.xml` file with these contents: https://gist.github.com/esr360/b03fc562102f49d1a696bf520d83fb16

It should be noted I only write CSS - I have no interest in Java, whilst I would love a separate front-end repo for my project it is not the case, so [seeing this](https://i.imgur.com/nT0tZCOl.png) makes me extremely uneasy as I have no idea what any of it is.

**Comment by snjeza** [comment]
>  I have no interest in Java

You can add the following property to your settings.json:

```
"java.import.maven.enabled": false
```
- remove the .project, .classpath files and the .settings directory
- call Java : Clean the Java language server workspace

**Comment by esr360** [comment]
@snjeza cheers for that. I'd prefer to see this be an opt-in feature - this has happened to me on at least two totally separate projects (in two different countries) so I find it difficult to believe there is good justification to having it enabled by default, unless my circumstances are just extremely unfortunate.

**Comment by fbricon** [comment]
We've pushed a tentative fix that should help prevent copying the content of the workspace to the bin/ directory when users delete .classpath. Let's see how it works in the real world, once vscode-java 0.49.0 is out (soon)

**Comment by anyflow** [comment]
I'm using vscode-java 0.50, but bin directory is still created...

> We've pushed a tentative fix that should help prevent copying the content of the workspace to the bin/ directory when users delete .classpath. Let's see how it works in the real world, once vscode-java 0.49.0 is out (soon)

**Comment by snjeza** [comment]
@anyflow could you provide a project example?

**Comment by hanct** [comment]
> Making it configurable would be great thing todo

I second this. Typically, i am using gradle to build the source codes, which will output to build folder. Now I see both folders, can easily get confuse.

**Comment by ppazos** [comment]
The annoying thing for me about this bin folder is the file search shows results from the bin, so if I look at the file name only, several times I open and start modifying the file under bin and not the real file. This takes extra time of debugging to know WTF is going on the open the right file from the project source.

**Comment by brimworks** [comment]
@fbricon this appears to still be an issue, can this be reopened?

**Comment by lackovic** [comment]
Just installed the plugin using Visual Studio Code `1.47.2`: the `./bin/` folder is still created upon project build.

Why was this issue closed? What was the resolution?

**Comment by shrey150** [comment]
This just happened to me as well on Visual Studio Code `1.47.3` when deleting `.classpath`, running extension version `0.65.0`. I've also encountered @ppazos 's issue regarding file search pulling from the `./bin/` folder. Could this be reopened and looked into?

**Comment by pan3793** [comment]
I found the folder `bin` is defined at  `.classpath`.`<classpathentry ...>`, `.classpath` is generated by this plugin automatically. Actually, I don't think `bin` is a good name for plugin temp folder, lots of projects use `bin` to store `script` files, so we can't roughly exclude `bin` in `.gitignore`. I wonder if there is possible to change the default folder name to widely used or some words start with `.`, like `temp`, `build`, `.bin`? 
At least, please provide a property to set it in `settings.json`.

**Comment by aweris** [comment]
Same here. We're using the `bin` folder for something else in the project. It would be great to override option for the name

**Comment by shua** [comment]
adding the following lines to our `build.gradle` worked for me (we have "build" set as our default output dir):
```
eclipse {
    classpath {
        defaultOutputDir = file('build/classes/java')
    }
}
```

**Comment by shua** [comment]
While I haven't found a way to change how vscode creates the initial config, I have found a hacky way to reconfigure the buildship plugin output paths. *Assuming* you are also using `build/` as your build output path (if you're using `target/` then edit the following accordingly):
```groovy
eclipse.classpath {
  defaultOutputDir = file("build")
  file.whenMerged { cp ->
    cp.entries.forEach { cpe ->
      if (cpe instanceof org.gradle.plugins.ide.eclipse.model.SourceFolder) {
        cpe.output = cpe.output.replace "bin/", "build/classes/java/"
      }
      if (cpe instanceof org.gradle.plugins.ide.eclipse.model.Output) {
        cpe.path = cpe.path.replace "bin/", "build/"
      }
    }
  }
}
```

**Comment by tigerinus** [comment]
Facing the same issue in 2021. Any update?

**Comment by khooz** [comment]
Just here to bump it up. Happened today when I was chopping a big project into smaller modules and moving stuff around. The `bin` suddenly popped up and reappears upon close -> delete -> reopening code. It happens if there's a `.project` file (an artifact of the past...) around. It's funny that this thread (which is an issue btw) is still active after more than 3 years.

**Comment by 0x416c69** [comment]
There's a .project for me popping up on 4 different directories in a mixed C++/Java project directory. (They're not in the same folder but in the same parent folder)

I've already disabled the "Autobuild"

I think an option which says "Don't fuck with my workspace in any way or form" would be very neat cuz I would enable it.

**Comment by dhkim09a** [comment]
I can't believe this is still unresolved after all these years. When I tried opening an Android OS source code directory with vscode, this plugin generated more than ten thousand unwanted files. Those files blocked building Android OS and overwrote version-controlled files. Imagine you check the version status of an OS source directory with hundreds git repositories and you find hundred git repositories are corrupted by thousands of unwanted files. It corrupts the directory very badly, and cleaning up generated files was really painful.

**Comment by idoybh** [comment]
> I can't believe this is still unresolved after all these years. When I tried opening an Android OS source code directory with vscode, this plugin generated more than ten thousand unwanted files. Those files blocked building Android OS and overwrote version-controlled files. Imagine you check the version status of an OS source directory with hundreds git repositories and you find hundred git repositories are corrupted by thousands of unwanted files. It corrupts the directory very badly, and cleaning up generated files was really painful.

`repo forall -c "git clean -f ."`

But yea I get why this is annoying. I have the same

**Comment by dhkim09a** [comment]
> > I can't believe this is still unresolved after all these years. When I tried opening an Android OS source code directory with vscode, this plugin generated more than ten thousand unwanted files. Those files blocked building Android OS and overwrote version-controlled files. Imagine you check the version status of an OS source directory with hundreds git repositories and you find hundred git repositories are corrupted by thousands of unwanted files. It corrupts the directory very badly, and cleaning up generated files was really painful.
> 
> `repo forall -c "git clean -f ."`
> 
> But yea I get why this is annoying. I have the same

Thank you. Actually, it's more than annoying when I have untracked changes. But still thank you for the suggestion.

**Comment by idoybh** [comment]
> Thank you. Actually, it's more than annoying when I have untracked changes. But still thank you for the suggestion.

Yea understandable. This plugin not working properly in AOSP is the only reason I'll keep using atom ;)

**Comment by ascertrobw** [comment]
> > Thank you. Actually, it's more than annoying when I have untracked changes. But still thank you for the suggestion.
> 
> Yea understandable. This plugin not working properly in AOSP is the only reason I'll keep using atom ;)

Agreed - happily use VSCode for many things, but my Java work remains on NB as a result of this one single issue.

**Comment by hakanai** [comment]
We had a working workaround for this which was just:

```
eclipse {
  classpath {
    defaultOutputDir = file('build/eclipse')
  }
}
```

At some point in the past few weeks, this stopped working and `bin/` directories started appearing again so now I'm experimenting with other options people have posted above.

It would be nice if VSCode would keep its junk to a `.vscode` directory or something instead of polluting directories which in some cases (particularly for a common name like `bin/`) we already have real files in.

**Comment by noordawod** [comment]
This plugin is creating a `bin` folder when opening a Gradle project. How do I configure it NOT to create such a folder? Or maybe direct it to `/dev/null`?

**Comment by arobertn** [comment]
Why is this issue closed? The referenced "fix" seems to be for Eclipse, not VS Code, and in any case seems not to be working judging by the continuing posts and upvotes. To that point I am getting this in a mixed Scala/Java Gradle project with "Language Support for Java by Red Hat 1.4.0" in VS Code 1.62.2.

**Comment by noordawod** [comment]
Seriously... Reopen?!

**Comment by darchangels13** [comment]
OH look, it's _still_ a thing.

**Comment by tim-hilt** [comment]
Not fixed for me! Can’t use it like this unfortunately

**Comment by yschimke** [comment]
This makes this unusable, is there some fix/workaround I'm missing?

**Comment by ryanrhee** [comment]
still happening.

**Comment by gardner** [comment]
This is still an issue. Why was this closed?

**Comment by chr7** [comment]
Still not fixed, please reopen!

**Comment by wch18735** [comment]
This just make me wanna use intellij again..

**Comment by fbricon** [comment]
If you guys feel strongly about this issue, please vote for https://github.com/gradle/gradle/issues/23032. We can't do anything until this is fixed upstream.

**Comment by tim-hilt** [comment]
This is an issue with gradle, however I also see the same behavior with Maven. How does the linked issue account for Maven?

**Comment by snjeza** [comment]
> This is an issue with gradle, however I also see the same behavior with Maven. How does the linked issue account for Maven?

@tim-hilt could you attach a project example reproducing the issue?

**Comment by gardner** [comment]
This does not occur when using gradle cli to build the project.
This does occur when using the vscode plugin.

**Comment by tim-hilt** [comment]
@snjeza unfortunately not atm, but I don't seem to be the only one seeing issues with Maven. Just search for "maven" on this page.

**Comment by snjeza** [comment]
@tim-hilt I can't reproduce the issue with Maven.

**Comment by tim-hilt** [comment]
I‘ll give it another try tomorrow and report back.

**Comment by RolKau** [comment]
This still happens in VSCode v1.74.3, Language support for Java v1.13.0. But it seems to be related to some internal state in the relevant workspaceStorage/xxx sub-directory. It would consistently generate a bin/ directory with a complete copy of the entire project, but after deleting the entire workspaceStorage/xxx directory and reopening the project, it would not, and the source code (notably the pom.xml and .project files) were exactly the same.

**Comment by gardner** [comment]
@snjeza They closed the issue. I uninstalled the extension.

This extension is not usable and the maintainers seem to have indicated they won't fix it. I understand there are a lot of competing priorities. I have moved on.

**Comment by sparrowt** [comment]
https://github.com/redhat-developer/vscode-java/issues/2344 is still open and (possibly) unblocked now, so here's hoping...

**Comment by MagnusMG** [comment]
My problem is that every so often I open a kotlin source file from the bin folder, without noticing that it is from the bin folder. And I make changes to that file. And three hours later I try to build the project, and all my changes are overwritten with the unchanged source file in the source folder.

I have disabled the extension now. It is a shame, because it is good apart from this.

**Comment by RolKau** [comment]
@sparrowt 
This will probably never be fixed, because the developers seems to believe that it is an issue with the Gradle backend, despite that it happens with Maven projects, too. There seems to be a timing issue and/or something written to the workspace storage that triggers it.

**Comment by darchangels13** [comment]
> @sparrowt This will probably never be fixed, because the developers seems to believe that it is an issue with the Gradle backend, despite that it happens with Maven projects, too. There seems to be a timing issue and/or something written to the workspace storage that triggers it.

It's the java language server that does it.  When VS Code parses a java project it generates these files.  You can improve the number of times this happens to you by turning off the Hybrid launch mode (otherwise you'll get it every time you start a workspace with open java projects) effectively degrading the performance of the editor.  But you won't get away from this unless you switch to the lightweight server entirely, which really reduces the usefulness since that's mostly good for syntax and nothing more.

They recycled the an Eclipse project for this.  Eclipse has a setting you can use to specify the output of  its temporary class files, and when specified to overlap with your build/target directory it's not nearly as painful an experience.  Unfortunately the VS Code extension doesn't seem to support configuration of this value and some combination of in-feasibility or intransigence has caused them pretend this is a non-issue that can't be reproduced or solved.

**Comment by RolKau** [comment]
> Eclipse has a setting you can use to specify the output of its temporary class files, and when specified to overlap with your build/target directory it's not nearly as painful an experience

But it is not just regular output files, it is like the extension suddenly forgets that it is a _Java_ language server and starts copying _all_ files.

**Comment by darchangels13** [comment]
> > Eclipse has a setting you can use to specify the output of its temporary class files, and when specified to overlap with your build/target directory it's not nearly as painful an experience
> 
> But it is not just regular output files, it is like the extension suddenly forgets that it is a _Java_ language server and starts copying _all_ files.

@RolKau I haven't seen that one, not yet, in any event.  I'm doing Gradle (+Spring) and it will very reliably build into that directory, a full set of class files, processed resources, etc.  no package artifacts for me.  It very much looks like the eclipse output from when I still used that beached whale of an editor.  

And this used to be something you'd have to correct in _**every**_ Eclipse project, but it was fine because, though it was obnoxious to do it, there was a setting and these settings would get checked in under the ".project" or one of the infinite number project bloatfiles eclipse generates.  And then it would disappear until the next time you were setting up a project.

And hey if i can make a .project file for this to go away, I will.  I'll actively resent it, but I'll do it.  But this really should be something the VS Code Extension can configure.  Nobody manages .class files in bin/ because the moment you're managing dependencies you've switched to one of several popular choices, **not one of which** uses bin/ for output.

**Comment by RolKau** [comment]
>  I'm doing Gradle (+Spring) and it will very reliably build into that directory, a full set of class files

For Maven it is far worse, because for each _POM_ project, it will copy everything that is not a .java or .class file into the bin/ subdirectory.

> these settings would get checked in under the ".project" or one of the infinite number project bloatfiles

I already have a .project file in each of the directories, but that said, your comment made me think that perhaps the problem is that there isn't any .classpath (because the POM projects don't have src or target), so I tried creating a minimal `.classpath` containing (non-directory fragments from a project at the time these files were generated in the source folder):

    <?xml version="1.0" encoding="UTF-8"?>
    <classpath>
        <classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-17">
            <attributes><attribute name="maven.pomderived" value="true"/></attributes>
        </classpathentry>
        <classpathentry kind="con" path="org.eclipse.m2e.MAVEN2_CLASSPATH_CONTAINER">
            <attributes><attribute name="maven.pomderived" value="true"/></attributes>
        </classpathentry>
    </classpath>

and although this seemed to cure the branch that was having the problem, it now also worked after deleting the .classpath file again. Given the mercurial nature of the problem, one can thus not be sure if this actually had an effect, or if the extension was just having a good day all along.

**Comment by ryeleo** [comment]
It seems this is closed, but I am facing this issue still using:

Visual Studio Code version: 1.82.0-insider
Java extension version: 1.22.1
JDK: openjdk 17.0.8.1 2023-08-24

**Comment by gardner** [comment]
@ryeleo you can leave a review at https://marketplace.visualstudio.com/items?itemName=redhat.java

Please consider leaving a comment to describe your star rating.

**Comment by ascertrobw** [comment]
> It seems this is closed, but I am facing this issue still. I am using:
> 
> Visual Studio Code version: 1.82.0-insider Java extension version: 1.22.1 JDK: openjdk 17.0.8.1 2023-08-24

Curious what Gradle version you are using? 

I've just got my main build up to Gradle 8.3 in the hope I'd be able to work around this - I spotted a note in Gradle 8.1 along the lines that the buildship/eclipse outputDir could now be explicitly configured. I literally have not been able to use vscode for the past 2 years or more due to this issue. Be great to be able to use it again if it's sorted. NetBeans is OK but there are times where I'd like to be able to use vscode for java project work.

**Comment by fbricon** [comment]
You guys can try the vscode-gradle extension with `"java.gradle.buildServer.enabled":"on"`. This contributes a new approach to Gradle support to vscode-java, that doesn't rely on Buildship, but delegates the build to the Gradle daemon instead. Doesn't generate a bin/ directory

**Comment by ascertrobw** [comment]
> You guys can try the vscode-gradle extension with `"java.gradle.buildServer.enabled":"on"`. This contributes a new approach to Gradle support to vscode-java, that doesn't rely on Buildship, but delegates the build to the Gradle daemon instead. Doesn't generate a bin/ directory

That is very good to know - will give it a try

**Comment by ascertrobw** [comment]
> vscode-gradle

Sorry - dumb Q - where would I set this? Not seeing it as an obvious setting on the RH pack

**Comment by ascertrobw** [comment]
> try the vscode-gradle extension with `"java.gradle.buildServer.enabled":"on"`

Sorry - dumb Q - where would I set this? Not seeing it as an obvious setting on the RH pack

**Comment by fbricon** [comment]
VSCode project or user settings.json

**Comment by fbricon** [comment]
The build language server won't touch already imported projects. Try the `Java: Clean the Java Language Server Workspace` command, to prune existing import state

**Comment by ascertrobw** [comment]
Will try that - but not sure it's the cause. This was a brand new clean project, cloned fresh from our repos. So there should have been no pre-existing project. Moment I opened it the bin folders started appearing. Suspect I have some other setting/extension wrong or missing.

**Comment by RolKau** [comment]
@ascertrobw 
What happens if you clone the same project into a _new_ directory and add the content below into a file called `.classpath` at the root project directory _before_ you open it in VSCode:

    <?xml version="1.0" encoding="UTF-8"?>
    <classpath>
        <classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-17"/>
        <classpathentry kind="con" path="org.eclipse.buildship.core.gradleclasspathcontainer"/>
    </classpath>

**Comment by ascertrobw** [comment]
Will have to make some time to try that next week

> @ascertrobw What happens if you clone the same project into a _new_ directory and add the content below into a file called `.classpath` at the root project directory _before_ you open it in VSCode:
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <classpath>
>     <classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-17"/>
>     <classpathentry kind="con" path="org.eclipse.buildship.core.gradleclasspathcontainer"/>
> </classpath>
> ```

Will see if I can make some time to try that next week - I presume that'll also work for Java 11? That's still our base build platform.

**Comment by RolKau** [comment]
> I presume that'll also work for Java 11?

My condolences! In that case you replace `JavaSE-17` with `JavaSE-11` at the end of the third line.

**Comment by jhirn** [comment]
> You guys can try the vscode-gradle extension with `"java.gradle.buildServer.enabled":"on"`. This contributes a new approach to Gradle support to vscode-java, that doesn't rely on Buildship, but delegates the build to the Gradle daemon instead. Doesn't generate a bin/ directory

Hey @fbricon!. This actually worked after running `Java: Clean the Java Language Server Workspace`. I'm not sure if there are other drawbacks to this but one thing I'm trying to get is classes in build to be updated on save for hot reload in a Spring Boot app. I am working around this now by running `gradle -t build` in a separate window, but it would be nice if VSCode were compiling these on save and updating `build/` to trigger a server reload. Is there some other setting I need to enable this?

**Comment by fbricon** [comment]
@jhirn no idea. Please ask on https://github.com/microsoft/vscode-gradle/issues

**Comment by adamsmd** [comment]
Here is a fix that works for me to prevent `bin` folders from automatically appearing: change `java.autobuild.enabled` to `false`.

(I don't know if `bin` folders will still appear if I manually build through VSCode, but at least it doesn't automatically appear just because I opened the editor.)

In the settings UI, this setting can be found under `Extensions: Language Support for Java(TM) by Red Hat` and is called `Java > AutoBuild: Enabled`.  Note that it can be set at either the user or the workspace level, depending on what you want.

**Comment by djkeh** [comment]
> Here is a fix that works for me to prevent `bin` folders from automatically appearing: change `java.autobuild.enabled` to `false`.
> 
> (I don't know if `bin` folders will still appear if I manually build through VSCode, but at least it doesn't automatically appear just because I opened the editor.)
> 
> In the settings UI, this setting can be found under `Extensions: Language Support for Java(TM) by Red Hat` and is called `Java > AutoBuild: Enabled`. Note that it can be set at either the user or the workspace level, depending on what you want.

This perfectly worked for me. I've been suffering for more than a couple of years not even knowing who's doing it. Many thanks @adamsmd !

**Comment by einarjohnson** [comment]
has this issue ever been fixed properly in vscode? facing the same issue here using Gradle in a multi-project setup. Super annoying.

**Comment by eleses** [comment]
I don't understand how this was "fixed" in 2020. It still does this nonsense for me in 2025.

Copies everything in source dir but the java files to java.project.outputPath. The java files it compiles though.

**Comment by eleses** [comment]
> change `java.autobuild.enabled` to `false`.

And what happens you click "Rebuild all"?! Full copy of everything but the java files to the outputPath.

**Comment by Nhangg** [comment]
> > You guys can try the vscode-gradle extension with `"java.gradle.buildServer.enabled":"on"`. This contributes a new approach to Gradle support to vscode-java, that doesn't rely on Buildship, but delegates the build to the Gradle daemon instead. Doesn't generate a bin/ directory
> 
> Hey [@fbricon](https://github.com/fbricon)!. This actually worked after running `Java: Clean the Java Language Server Workspace`. I'm not sure if there are other drawbacks to this but one thing I'm trying to get is classes in build to be updated on save for hot reload in a Spring Boot app. I am working around this now by running `gradle -t build` in a separate window, but it would be nice if VSCode were compiling these on save and updating `build/` to trigger a server reload. Is there some other setting I need to enable this?

this saved me as well. took hours to find work-around way.

**Comment by slowhigh** [comment]
The best approach is to create the file below and configure it as shown to hide the folders.

File path: `.vscode/settings.json`

```json
{
    "files.exclude": {
        "**/build": true,
        "**/bin": true
    }
}
```

**Comment by awecz** [comment]
This nonsense is still present in 2026.
